### PR TITLE
Move Statsbeat to Exporter Module

### PR DESCRIPTION
**Pull request by heyams**

### CODE REVIEW

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 1]
```diff
@@ -197,16 +197,20 @@ public void customize(AutoConfigurationCustomizer autoConfiguration) {
...
-          initStatsbeatTelemetryItemExporter(
-              statsbeatModule, tempDir, configuration.preview.diskPersistenceMaxSizeMb),
-          telemetryClient::getStatsbeatConnectionString,
-          telemetryClient::getInstrumentationKey,
-          configuration.internal.statsbeat.disabledAll,
-          configuration.internal.statsbeat.shortIntervalSeconds,
-          configuration.internal.statsbeat.longIntervalSeconds,
-          configuration.preview.statsbeat.disabled,
-          initStatsbeatFeatureSet(configuration));
+      try {
```
Comment (heyams) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:200]: @trask we want to wait for your input on this try/catch block.

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 2]
```diff
@@ -248,6 +271,140 @@ public void customize(AutoConfigurationCustomizer autoConfiguration) {
...
+              diskPersistenceMaxSizeMb,
+              TempDirs.getSubDir(tempDir, STATSBEAT_FOLDER_NAME),
+              telemetryPipeline,
+              LocalStorageStats.noop(),
+              true);
+      telemetryPipelineListener =
+          TelemetryPipelineListener.composite(
+              new StatsbeatTelemetryPipelineListener(
+                  () -> {
+                    statsbeatModule.shutdown();
```
Comment (trask) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:294]: my guess is `shutdown` is being called somehow, which is terminating the ExecutorService

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 3]
```diff
@@ -184,7 +196,18 @@ public void customize(AutoConfigurationCustomizer autoConfiguration) {
     }
 
     // initialize StatsbeatModule
-    statsbeatModule.start(telemetryClient, configuration);
+    if (telemetryClient.getConnectionString() != null) {
```
Comment (jeanbisutti) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:199]: Can we add a comment on the reason for this check?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 4]
```diff
@@ -248,6 +271,140 @@ public void customize(AutoConfigurationCustomizer autoConfiguration) {
         .addShutdownHook(new Thread(() -> flushAll(telemetryClient).join(10, TimeUnit.SECONDS)));
   }
 
+  private static TelemetryItemExporter initStatsbeatTelemetryItemExporter(
```
Comment (jeanbisutti) [inline agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:274]: Should this method be migrated into the export to allow stat beat to use the file system?

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Feature.java [hunk 5]
```diff
@@ -1,14 +1,14 @@
...
-package com.microsoft.applicationinsights.agent.internal.statsbeat;
+package com.azure.monitor.opentelemetry.exporter.implementation.statsbeat;
 
 import java.util.BitSet;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Set;
 
-enum Feature {
+public enum Feature {
```
Comment (jeanbisutti) [inline agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Feature.java:11]: Should it be public, that is to say i.e. accessible to users of the standalone exporter?

File: agent/azure-monitor-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/StatsbeatTestUtils.java [hunk 6]
```diff
@@ -156,5 +182,74 @@ private static <E> Set<E> decode(
     return result;
   }
 
+  static TelemetryItemExporter getTelemetryItemExporter() {
```
Comment (jeanbisutti) [inline agent/azure-monitor-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/StatsbeatTestUtils.java:185]: This method does not seem to be used

**Review by jeanbisutti — COMMENTED** [review]
Would it be possible to add a test in the standalone exporter showing how to use stats beats in the standalone exporter (I imagine with `statsbeatModule.start`)?

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java [hunk 7]
```diff
@@ -248,6 +271,140 @@ public void customize(AutoConfigurationCustomizer autoConfiguration) {
...
+                    statsbeatModule.shutdown();
+                    localStorageTelemetryPipelineListener.shutdown();
+                  }),
+              localStorageTelemetryPipelineListener);
+    }
+
+    return new TelemetryItemExporter(telemetryPipeline, telemetryPipelineListener);
+  }
+
+  private static Set<Feature> initStatsbeatFeatureSet(Configuration config) {
```
[the remark this answers is not in the document]
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:303]: same

File: agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java (hunk 3 above)
Comment (heyams) [reply agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/SecondEntryPoint.java:199]: we don't want to start statsbeat when connection string is null. we have this check in many places.

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/FeatureStatsbeat.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+  long[] getInstrumentation() {
+    return Instrumentations.encode(instrumentationList);
+  }
+
+  // this is used by Exporter
+  public void addInstrumentation(String instrumentation) {
+    instrumentationList.add(instrumentation);
+  }
+
+  // this is used by ByteCodeUtil
```
[the remark this answers is not in the document]
Comment (heyams) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/FeatureStatsbeat.java:47]: true, i can move it out of exporter.

File: agent/azure-monitor-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/StatsbeatTestUtils.java (hunk 6 above)
Comment (heyams) [reply agent/azure-monitor-exporter/src/test/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/StatsbeatTestUtils.java]: removed.

File: agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Feature.java (hunk 5 above)
Comment (heyams) [reply agent/azure-monitor-exporter/src/main/java/com/azure/monitor/opentelemetry/exporter/implementation/statsbeat/Feature.java:13]: add a todo

**Review by jeanbisutti — APPROVED** [review]

**Pull request merged**